### ISSUE TITLE
Failed to run GC

### ISSUE THREAD

**Issue by hlwanghl** [goharbor/harbor#9044]
**Expected behavior and actual behavior:**
Expected: successful to run GC.
Actual: Failed to run GC.

**Steps to reproduce the problem:**
1. push some images.
2. delete some images through Harbor web console.
3. run GC through Harbor web console.
4. Harbor reported `Failed to start GC: 500`.

**Versions:**

- harbor version: [1.8.2]
- docker engine version: [18.09.6]
- docker-compose version: [1.24.0]

**Additional context:**

GC with the same configuration works for Harbor 1.7.5 and below, but not 1.8.x.

- **Harbor config files:**
```
storage_service:
  s3:
    accesskey: xxx
    secretkey: yyy
    region: us-west-1
    regionendpoint: https://s3.pek3a.qingstor.com
    bucket: harbor2
    encrypt: true
    secure: true
    v4auth: true
    chunksize: 5242880
    multipartcopychunksize: 33554432
    multipartcopymaxconcurrency: 100
    multipartcopythresholdsize: 33554432
    rootdirectory: /v180
```
- **Log files:**
```
==> core.log <==
Sep 10 20:03:23 bogon core[975]: 2019/09/10 12:03:23 #033[1;44m[D] [server.go:2619] |192.168.255.254|#033[42m 201 #033[0m|   6.543687ms|   match|#033[46m POST    #033[0m /api/system/gc/schedule   r:/api/system/gc/schedule#033[0m

==> proxy.log <==
Sep 10 20:03:23 bogon proxy[975]: 192.168.255.254 - "POST /api/system/gc/schedule HTTP/1.1" 201 0 "http://192.168.2.34/harbor/gc" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36" 0.008 0.008 .

==> jobservice.log <==
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/jobservice/worker/cworker/c_worker.go:70]: Job incoming: {"name":"IMAGE_GC","id":"72563a8dfb721719125d2d1b","t":1568117003,"args":{"redis_url_reg":"redis://redis:6379/1"},"unique":true}
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/common/config/store/driver/rest.go:24]: get configuration from url: http://core:8080/api/internal/configurations

==> core.log <==
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   1.609741ms|   match|#033[44m GET     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m

==> jobservice.log <==
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/common/config/store/driver/rest.go:24]: get configuration from url: http://core:8080/api/internal/configurations

==> core.log <==
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   1.163097ms|   match|#033[44m GET     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m

==> jobservice.log <==
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/common/registryctl/client.go:41]: initializing client for reigstry http://registryctl:8080 ...
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/common/config/store/driver/rest.go:24]: get configuration from url: http://core:8080/api/internal/configurations

==> core.log <==
Sep 10 20:03:24 bogon core[975]: 2019-09-10T12:03:24Z [INFO] [/core/service/notifications/admin/handler.go:82]: received admin job status update event: job-39, status-running
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   4.263248ms|   match|#033[46m POST    #033[0m /service/notifications/jobs/adminjob/39   r:/service/notifications/jobs/adminjob/:id([0-9]+)#033[0m
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   11.99874ms|   match|#033[44m GET     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   3.822184ms|   match|#033[43m PUT     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|  28.483301ms|   match|#033[43m PUT     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m

==> jobservice.log <==
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [INFO] [/jobservice/runner/redis.go:161]: start to run gc in job.

==> registryctl.log <==
Sep 10 20:03:24 bogon registryctl[975]: 2019-09-10T12:03:24Z [ERROR] [/registryctl/api/registry.go:49]: Fail to execute GC: exit status 2, command err: panic: runtime error: invalid memory address or nil pointer dereference
Sep 10 20:03:24 bogon registryctl[975]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc51d8d]
Sep 10 20:03:24 bogon registryctl[975]:
Sep 10 20:03:24 bogon registryctl[975]: goroutine 1 [running]:
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).doWalk.func1(0xc000389b80, 0xc000399a01, 0xc0001e16e8)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:973 +0x9d
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/vendor/github.com/aws/aws-sdk-go/service/s3.(*S3).ListObjectsV2PagesWithContext(0xc00000c930, 0x7f1ed4ada188, 0xc00044ef50, 0xc000296d20, 0xc0004f97e8, 0x0, 0x0, 0x0, 0x1, 0x2)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/vendor/github.com/aws/aws-sdk-go/service/s3/api.go:4198 +0x111
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).doWalk(0xc000389b00, 0xff4d80, 0xc00044eee0, 0xc0001e18a8, 0xc000374cf1, 0x25, 0x0, 0x0, 0xc000431900, 0x0, ...)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:971 +0x3a1
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).Walk(0xc000389b00, 0xff4d80, 0xc00044eee0, 0xc00044c940, 0x20, 0xc000431900, 0x2, 0x0)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:919 +0x160
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage/driver/base.(*Base).Walk(0xc00042d0d0, 0xff4d80, 0xc00044eee0, 0xc00044c940, 0x20, 0xc000431900, 0x0, 0x0)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go:239 +0x234
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage.(*registry).Enumerate(0xc00044ee70, 0xff4ac0, 0xc00045a360, 0xc0003999c0, 0x0, 0x0)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/catalog.go:67 +0xe3
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry/storage.MarkAndSweep(0xff4ac0, 0xc00045a360, 0x1000c20, 0xc00042d0d0, 0xff6c80, 0xc00044ee70, 0x100, 0xff6c80, 0xc00044ee70)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/storage/garbagecollect.go:40 +0x1e7
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/registry.glob..func3(0x171eca0, 0xc0004303c0, 0x1, 0x2)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/registry/root.go:80 +0x48a
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).execute(0x171eca0, 0xc0004302e0, 0x2, 0x2, 0x171eca0, 0xc0004302e0)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:495 +0x191
Sep 10 20:03:24 bogon registryctl[975]: github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).Execute(0x171efe0, 0xc0001e1f88, 0xc00008c058)
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:560 +0x2f4
Sep 10 20:03:24 bogon registryctl[975]: main.main()
Sep 10 20:03:24 bogon registryctl[975]: #011/go/src/github.com/docker/distribution/cmd/registry/main.go:23 +0x2d
Sep 10 20:03:24 bogon registryctl[975]: 172.20.0.9 - - [10/Sep/2019:12:03:24 +0000] "POST /api/registry/gc HTTP/1.1" 500 22

==> core.log <==
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   3.678313ms|   match|#033[43m PUT     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|  31.089735ms|   match|#033[43m PUT     #033[0m /api/internal/configurations   r:/api/internal/configurations#033[0m

==> jobservice.log <==
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [ERROR] [/registryctl/client/client.go:94]: Failed to start gc: 500
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [ERROR] [/jobservice/runner/redis.go:161]: failed to get gc result: Failed to start GC: 500
Sep 10 20:03:24 bogon jobservice[975]: 2019-09-10T12:03:24Z [ERROR] [/jobservice/runner/redis.go:63]: |@_@| Job 'IMAGE_GC:72563a8dfb721719125d2d1b' exit with error: Failed to start GC: 500

==> core.log <==
Sep 10 20:03:24 bogon core[975]: 2019-09-10T12:03:24Z [INFO] [/core/service/notifications/admin/handler.go:82]: received admin job status update event: job-39, status-error
Sep 10 20:03:24 bogon core[975]: 2019/09/10 12:03:24 #033[1;44m[D] [server.go:2619] |     172.20.0.9|#033[42m 200 #033[0m|   2.841207ms|   match|#033[46m POST    #033[0m /service/notifications/jobs/adminjob/39   r:/service/notifications/jobs/adminjob/:id([0-9]+)#033[0m
```
- **Screenshots:**
![image](https://user-images.githubusercontent.com/4861515/64612445-6885c400-d406-11e9-99e1-bfb1f8dd0ca5.png)

**Comment by wy65701436** [comment]
Can you try to execute the GC command in the registry container?

**Comment by hlwanghl** [comment]
@wy65701436 thanks for reply. The gc command reported the same error:

```
root@i-f9kktsvu:~/harbor# docker exec -u harbor registry bin/registry garbage-collect /etc/registry/config.yml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc51d8d]

goroutine 1 [running]:
github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).doWalk.func1(0xc000327980, 0xc000430801, 0xc0001d96e8)
	/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:973 +0x9d
github.com/docker/distribution/vendor/github.com/aws/aws-sdk-go/service/s3.(*S3).ListObjectsV2PagesWithContext(0xc00008e808, 0x7fe4355395e8, 0xc000440ee0, 0xc000292eb0, 0xc0004b37e8, 0x0, 0x0, 0x0, 0x1, 0x2)
	/go/src/github.com/docker/distribution/vendor/github.com/aws/aws-sdk-go/service/s3/api.go:4198 +0x111
github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).doWalk(0xc000327900, 0xff4d80, 0xc000440e70, 0xc0001d98a8, 0xc000402901, 0x25, 0x0, 0x0, 0xc000448c20, 0x0, ...)
	/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:971 +0x3a1
github.com/docker/distribution/registry/storage/driver/s3-aws.(*driver).Walk(0xc000327900, 0xff4d80, 0xc000440e70, 0xc00043e5e0, 0x20, 0xc000448c20, 0x2, 0x0)
	/go/src/github.com/docker/distribution/registry/storage/driver/s3-aws/s3.go:919 +0x160
github.com/docker/distribution/registry/storage/driver/base.(*Base).Walk(0xc0004427b0, 0xff4d80, 0xc000440e70, 0xc00043e5e0, 0x20, 0xc000448c20, 0x0, 0x0)
	/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go:239 +0x234
github.com/docker/distribution/registry/storage.(*registry).Enumerate(0xc000440e00, 0xff4ac0, 0xc000426f90, 0xc000430780, 0x0, 0x0)
	/go/src/github.com/docker/distribution/registry/storage/catalog.go:67 +0xe3
github.com/docker/distribution/registry/storage.MarkAndSweep(0xff4ac0, 0xc000426f90, 0x1000c20, 0xc0004427b0, 0xff6c80, 0xc000440e00, 0x0, 0xff6c80, 0xc000440e00)
	/go/src/github.com/docker/distribution/registry/storage/garbagecollect.go:40 +0x1e7
github.com/docker/distribution/registry.glob..func3(0x171eca0, 0xc0003199c0, 0x1, 0x1)
	/go/src/github.com/docker/distribution/registry/root.go:80 +0x48a
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).execute(0x171eca0, 0xc000319970, 0x1, 0x1, 0x171eca0, 0xc000319970)
	/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:495 +0x191
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).Execute(0x171efe0, 0xc0001d9f88, 0xc000076058)
	/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:560 +0x2f4
main.main()
	/go/src/github.com/docker/distribution/cmd/registry/main.go:23 +0x2d
root@i-f9kktsvu:~/harbor#
```

**Comment by hlwanghl** [comment]
Forgot to mention we are using [QingStor](https://docs.qingcloud.com/qingstor/s3/) as S3-compatible backend.

**Comment by bitsf** [comment]
I think we should track the registry version that we're using in the release, or it's hard to check whether such issue is introduced by upstream registry or harbor itself

**Comment by hlwanghl** [comment]
any updates? is it possible to use an older version of docker registry because it worked in Harbor 1.7.x? e.g. Harbor 1.8.0 with goharbor/registry-photon:v2.6.2-v1.7.6.

**Comment by bitsf** [comment]
I think you can have a try, we haven't tested on QingStor as S3-compatible backend, that can help narrow the scope of the problem.

**Comment by wy65701436** [comment]
Hi, @hlwanghl , you could find the foot cause at https://github.com/goharbor/harbor/issues/9186. And we have already fixed it it v1.9.1-rc1

**Comment by hlwanghl** [comment]
> Hi, @hlwanghl , you could find the foot cause at #9186. And we have already fixed it it v1.9.1-rc1

It works now. Looking forward to 1.9.1 Release. Thanks!
